File: typescript/resistor-color-trio/resistor-color-trio.ts
const RESISTOR_COLORS: readonly string[] = [
  'black', 'brown', 'red', 'orange', 'yellow', 'green', 'blue', 'violet', 'grey', 'white',
] as const;

type Color = typeof RESISTOR_COLORS[number];

const colorCode = (color: Color): number => RESISTOR_COLORS.indexOf(color);

export function decodedResistorValue(colors: Color[]): string {
  if (colors.length < 3) {
    throw new Error('"colors" must contain at least 3 colors.');
  }

  const baseValue = 10 * colorCode(colors[0]) + colorCode(colors[1]);
  const multiplier = Math.pow(10, colorCode(colors[2]));
  const value = baseValue * multiplier;

  const metricPrefixes: readonly string[] = ['', 'kilo', 'mega', 'giga'];
  for (let index = metricPrefixes.length - 1; index >= 0; index--) {
    const threshold = Math.pow(1000, index);
    if (value >= threshold) {
      return `${value/threshold} ${metricPrefixes[index]}ohms`;
    }
  }

  return '0 ohms';
}

File: java/resistor-color-trio/src/main/java/ResistorColorTrio.java
import java.util.List;

public class ResistorColorTrio {
    String label(String[] colors) {
        return new Resistance(value(colors)).toString();
    }

    private static long value(String[] colors) {
        return (10 * colorCode(colors[0]) + colorCode(colors[1])) * (long) Math.pow(10, colorCode(colors[2]));
    }

    private static final List<String> COLORS = List.of("black", "brown", "red", "orange", "yellow", "green", "blue",
            "violet", "grey", "white");

    private static int colorCode(String color) {
        return COLORS.indexOf(color);
    }
}
